### ISSUE TITLE
Error in assocTestSingle() results when covariate is confounded with SNP genotype.

### ISSUE THREAD

**Issue by rmsegnitz** [smgogarten/GENESIS#87]
Hello and thanks for your great package which I've been enjoying using!

I have encountered a bug that seems to result when a covariate provided to GENESIS is perfectly confounded with one or more 2-genotype snps in the dataset. This seems likely only to occur in small datasets then the confounding occurs by chance; in our case we have a small cohort of around 50 individuals. I found that SNPs which were confounded with sex (included as a covariate) were found to be highly significant (p values approaching 0) even when there are no true association. I think that perhaps this could be caused by handling of NAs that result when trying to fit a perfectly confounded covariate, though I haven't dug into the package's internal functions to try to find where this arises.

I've reproduced this problem [here](https://github.com/rmsegnitz/GENESIS_troubleshoot/blob/main/1_Bug_Example.Rmd) using data available from SNPRelate. I also compared GENESIS results to models fit using lm() to highlight where confounded SNPs are clearly estimated incorrectly.

Thanks and I hope this is helpful!

**Comment by rmsegnitz** [comment]
In my analysis I encountered such SNPs as extreme outliers similar to the SNP demonstrated in the linked script. Here a SNP that is confounded with a covariate results in an extreme outlier  &  large deviation from corresponding lm() fit. 

![image](https://user-images.githubusercontent.com/43895313/155434456-0fd9c853-421a-44a1-9d33-c5dfce587622.png)

![image](https://user-images.githubusercontent.com/43895313/155434489-470d9768-c1d4-4b8a-bc95-f6d30308c5e1.png)

**Comment by amstilp** [comment]
We have also encountered this behavior before in GENESIS, when running a conditional analysis with SNPs that are perfectly correlated with one of the conditional variants. I agree it would be ideal to try to detect this and set them to NA it in the association testing, but we don't have the bandwidth to try to fix it at the moment. For now, we suggest filtering your association test results to remove any variants where the absolute value of the estimated effect and standard error is above some threshold (e.g., >1e10).

I'll leave the issue open in case we do get a chance to make an update to address this.

**Comment by rmsegnitz** [comment]
Great, just wanted to make sure developers were aware! It has of course been easy enough to simply filter the problematic results. Thanks!

**Comment by amstilp** [comment]
Yes, no problem! Thanks for the report. Also useful for other users who might run into the same behavior.

**Comment by HuaLing1** [comment]
I've came across the same issue while doing conditional analyses on the same SNP. However, what's really intriguing is it only happens to the snp with MAF of 0.005. While i was doing similar things on some common SNPs, this did not happen. Is this just coincidence or there is a reason behind?

**Comment by mconomos** [comment]
This happens because of the way that GENESIS uses two steps to perform the association testing: (1) fit the regression model with `fitNullModel` -- this is where you've included your covariates; (2) perform score tests for each variant with `assocTestSingle`. If you run a score test for a variant in step 2 after conditioning on a covariate in step 1 that is perfectly collinear, the results of the score test blow up. You shouldn't be able to get a result of that score test, but as mentioned, we don't currently have the bandwidth to implement a fix. 

I wanted to make a note about detecting this. As depicted in the original issue, this can often be seen as a very large effect size estimate `Est`. However, what is actually happening is that the score value (`Score`) and its standard error (`Score.SE`) are both very close to 0 (e.g. < 1e-10). When `Score` and `Score.SE` are both approximately 0, the effect size estimate (`Est`) and the test statistic (`Score.Stat`) are both essentially computed as the ratio 0/0. Depending on the actual values of the score and its se, these ratios can get very big or very small, and consequentially, the p-value for that variant can be either highly significant or very insignificant (i.e. p approximately 1). Therefore, the best way to detect this is to check for variants where _both_ the absolute value of the `Score` and the `Score.SE` are very small.
